### PR TITLE
Propose presentation

### PR DESCRIPTION
**Pull request by michaeldboyd**
#### Short description of what this resolves:
Implemented the [Propose Proof](https://github.com/hyperledger/aries-rfcs/tree/master/features/0037-present-proof#propose-presentation) message exchange of the Present Proof protocol

#### Changes proposed in this pull request:

- Added Create and Process messages in the DefaultProofService
- Added new message types and protocol states
- Added unit tests for better code coverage

I've also made 2 (hopefully pretty lightweight) additions in the Presentation Preview message format:
- Added schema_id and issuer_did to each `attribute` and `predicate` object in the `presentation_proposal` message (which is of type `presentation-preview`
- Added a `referent` to the predicate object to identify it in the proof request.

Here's the updated message format: 

```
{
    "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/present-proof/1.0/propose-presentation",
    "@id": "<uuid-propose-presentation>",
    "comment": "some comment",
    "presentation_proposal":
    {
        "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/present-proof/1.0/presentation- 
   preview",
        "attributes": [
            {
                "name": "<attribute_name>",
                "cred_def_id": "<cred_def_id>",
                "schema_id": "<schema_id>"    // NEW
                "issuer_did": "<issuer_did>",      // NEW
                "mime-type": "<type>",
                "value": "<value>",
                "referent": "<referent>"
            },
            // more attributes
        ],
        "predicates": [
            {
                "name": "<attribute_name>",
                "cred_def_id": "<cred_def_id>",
                "schema_id": "<schema_id>",    // NEW
                "issuer_did": "<issuer_did>"      // NEW
                "predicate": "<predicate>",
                "threshold": <threshold>,
                "referent" : "<referent>"         // NEW
            },
            // more predicates
        ]
    }
}

```
Some considerations:
- all attributes that share the same referent must also match the same cred_def_id, schema_id, and issuer_id restrictions.
- If these *id fields are left empty, the value will be self-attested
- the predicate referents must be unique within the list of referents. We have included them here so that an implementor can specify the naming convention of the referent. This is because the Trinsic wallet currently uses `referent` as a title for each Attribute or Predicate proof. We recognize this is likely not ideal, and we will likely move to a more unique local identifier in the future. 
- the `value` field is currently ignored, as there is no way for the requestor to specify `revealed` or `hidden` in the proof request they send back. The holder can choose whether they send an attribute as revealed during the presentation exchange and the verifier can decide if that fulfills their requirements at that point.

### CODE REVIEW

File: src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        /// <summary>
+        /// Gets or sets the proposed predicates.
+        /// </summary>
+        /// <value>The proposed predicates.</value>
+        [JsonProperty("predicates", NullValueHandling = NullValueHandling.Ignore)]
+        public List<ProposedPredicate> ProposedPredicates { get; set; } = new List<ProposedPredicate>();
+
+
+        /// <inheritdoc />
+        public override string ToString() =>
```
Comment (tmarkovski) [inline src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs:36]: Let's remove these `ToString` overrides.

File: test/Hyperledger.Aries.Tests/tester [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#! /bin/zsh
```
Comment (tmarkovski) [inline test/Hyperledger.Aries.Tests/tester:1]: Move this script to `scripts` folder instead? Maybe at `.sh` extension to it?

File: src/Hyperledger.Aries/Features/PresentProof/DefaultProofService.cs [hunk 3]
```diff
@@ -386,6 +386,255 @@ private async Task<string> BuildCredentialDefinitionsAsync(IAgentContext agentCo
...
+            var proofRecord = new ProofRecord
+            {
+                Id = Guid.NewGuid().ToString(),
+                ProposalJson = proposalJson,
+                ConnectionId = connection?.Id,
+                State = ProofState.Proposed
+            };
+
+            proofRecord.SetTag(TagConstants.LastThreadId, proposePresentationMessage.GetThreadId());
+            proofRecord.SetTag(TagConstants.Role, TagConstants.Requestor);
```
[the remark this answers is not in the document]
Comment (michaeldboyd) [reply src/Hyperledger.Aries/Features/PresentProof/DefaultProofService.cs:499]: Isn't the verifier the one processing and then requesting the proposal that the holder has just sent to them? I've made the verifier the `requestor`, as  they are requesting verification of the proposed proof. I agree that it is confusing with `requestor` in the sense that the holder is actually the initiator, or the `proposer`. Perhaps we should revisit this in a future PR

File: src/Hyperledger.Aries/Features/PresentProof/IProofService.cs [hunk 4]
```diff
@@ -12,6 +12,44 @@ namespace Hyperledger.Aries.Features.PresentProof
     /// </summary>
     public interface IProofService
     {
+
+        /// <summary>
+        /// Creates and sends a proof proposal to a connection.
```
[the remark this answers is not in the document]
Comment (michaeldboyd) [reply src/Hyperledger.Aries/Features/PresentProof/IProofService.cs:17]: Changed. I've also removed ""and sends" from the CreateCredentialAsync method summaries , as their logic is the same as these.

File: src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs (hunk 1 above)
Comment (michaeldboyd) [reply src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs:36]: These seem useful for both me and other users. Why remove them?

File: test/Hyperledger.Aries.Tests/tester (hunk 2 above)
Comment (michaeldboyd) [reply test/Hyperledger.Aries.Tests/tester:1]: Done

File: src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs (hunk 1 above)
Comment (michaeldboyd) [reply src/Hyperledger.Aries/Features/PresentProof/Models/ProofProposal.cs]: I've removed them, but still curious

**Comment by tmarkovski** [comment]
Created #102 to track record direction

**Pull request merged**